pkg/k8s/statedb.go

```diff
@@ -0,0 +1,329 @@
+	}
+
+	queryAll := r.QueryAll
+	if queryAll == nil {
+		// No query function provided, use All()
+		queryAll = QueryAllFunc[Obj](func(txn statedb.ReadTxn, tbl statedb.Table[Obj]) statedb.Iterator[Obj] {
+			return tbl.All(txn)
+		})
+	}
+
+	// Construct a stream of K8s objects, buffered into chunks every [waitTime] period
+	// and then committed.
+	// This reduces the number of write transactions required and thus the number of times
+	// readers get woken up, which results in much better overall throughput.
+	src := stream.Buffer(
+		ListerWatcherToObservable(r.ListerWatcher),
+		bufferSize,
+		waitTime,
+
+		// Buffer the events into a map, coalescing them by key.
+		func(buf *buffer, ev CacheStoreEvent) *buffer {
+			switch {
+			case ev.Kind == CacheStoreEventReplace:
+				return &buffer{
+					replaceItems: ev.Obj.([]any),
+					entries:      make(map[string]entry, bufferSize), // Forget prior entries
+				}
+			case buf == nil:
+				buf = &buffer{
+					replaceItems: nil,
+					entries:      make(map[string]entry, bufferSize),
+				}
+			}
+
+			var entry entry
+			entry.deleted = ev.Kind == CacheStoreEventDelete
+
+			var key string
+			if d, ok := ev.Obj.(cache.DeletedFinalStateUnknown); ok {
+				key = d.Key
+				var err error
+				entry.namespace, entry.name, err = cache.SplitMetaNamespaceKey(d.Key)
+				if err != nil {
+					panic(fmt.Sprintf("%T internal error: cache.SplitMetaNamespaceKey(%q) failed: %s", r, d.Key, err))
+				}
+				entry.obj, ok = transform(d.Obj)
+				if !ok {
+					return buf
+				}
+			} else {
+				meta, err := meta.Accessor(ev.Obj)
+				if err != nil {
+					panic(fmt.Sprintf("%T internal error: meta.Accessor failed: %s", r, err))
+				}
+				entry.name = meta.GetName()
+				if ns := meta.GetNamespace(); ns != "" {
+					key = ns + "/" + meta.GetName()
+					entry.namespace = ns
+				} else {
+					key = meta.GetName()
+				}
+
+				var ok bool
+				entry.obj, ok = transform(ev.Obj)
+				if !ok {
+					return buf
+				}
+			}
+			buf.entries[key] = entry
+			return buf
+		},
+	)
+
+	commitBuffer := func(buf *buffer) {
+		txn := r.db.WriteTxn(table)
+		defer txn.Commit()
+
+		numInserted, numDeleted := 0, 0
+		if buf.replaceItems != nil {
+			iter := queryAll(txn, table)
+			for obj, _, ok := iter.Next(); ok; obj, _, ok = iter.Next() {
+				numDeleted++
+				table.Delete(txn, obj)
+			}
+			for _, item := range buf.replaceItems {
+				if obj, ok := transform(item); ok {
+					table.Insert(txn, obj)
+					numInserted++
+				}
+			}
+			// Mark the table as initialized. Internally this has a sync.Once
+			// so safe to call multiple times.
+			r.initDone(txn)
+		}
+
+		for _, entry := range buf.entries {
+			if !entry.deleted {
+				numInserted++
+				table.Insert(txn, entry.obj)
+			} else {
+				table.Delete(txn, entry.obj)
+				numDeleted++
+			}
+		}
+
+		numTotal := table.NumObjects(txn)
+		health.OK(fmt.Sprintf("%d inserted, %d deleted, %d total", numInserted, numDeleted, numTotal))
```

Comment:
I'm wondering if it wouldn't be more logical to report this after `txn.Commit()`?

Reply:
Sure we can do that, though it doesn't make too much of a difference in practice. But good practice to do fever things within the WriteTxn.